### PR TITLE
Fix an issue in restoring ParallelMapIterator and add tests for ParallelMapDataset 

### PR DESCRIPTION
**Pull request by feihugis**
This PR fixes two issues and adds tests for `ParallelMapDataset`:

- https://github.com/tensorflow/tensorflow/commit/ecece7f8976c4da355443d998f5773a0598314a9: Clear the invocation results in the buffer before restoring `ParallelMapIterator`

- https://github.com/tensorflow/tensorflow/commit/098d891c6d8f3d9d65287efd27a2b47b304b8217: Add tests for `ParallelMapDataset`

- https://github.com/tensorflow/tensorflow/commit/f254f5c1c4b76287866a68d607664eccead1551f: Change the name of `thread_pool_` from `inter_op` to `intra_op_thread_op`. 

cc: @jsimsa

### CODE REVIEW

File: tensorflow/core/kernels/data/dataset_test_base.cc [hunk 1]
```diff
@@ -202,7 +202,7 @@ Status DatasetOpsTestBase::InitThreadPool(int thread_num) {
         "The `thread_num` argument should be positive but got: ", thread_num);
   }
   thread_pool_ = absl::make_unique<thread::ThreadPool>(
-      Env::Default(), ThreadOptions(), "inter_op", thread_num);
+      Env::Default(), ThreadOptions(), "intra_op_thread_pool", thread_num);
```
Comment (jsimsa) [inline tensorflow/core/kernels/data/dataset_test_base.cc]: call this simply "test_thread_pool"

**Review by jsimsa — CHANGES_REQUESTED** [review]

**Comment by feihugis** [comment]
@jsimsa 

When testing `ParallelMapDataset`, there is a timeout issue for [TestCase7](https://github.com/tensorflow/tensorflow/commit/098d891c6d8f3d9d65287efd27a2b47b304b8217#diff-d1771146ec6106d05f61244c9c8baa81R220). The detailed log is posted [here](https://gist.github.com/feihugis/7c9153d75efd5a000026a7e33a036a39).

It seems to be caused by that `num_parallel_calls` is too big and `thread_num` of `thread_pool_` is small. Note that the timeout case uses `use_inter_op_parallelism=false`.

- When `num_parallel_calls=2` and `thread_num=2`, the test times out;

- When `num_parallel_calls=1` and `thread_num=2`, the test works fine.

I tried debugging to figure out why it happens. It seems to be related to the executor running multiple frames asynchronously, but I did not find the root reason. Do you have any ideas or suggestions?

File: tensorflow/core/kernels/data/dataset_test_base.cc (hunk 1 above)
Comment (feihugis) [reply tensorflow/core/kernels/data/dataset_test_base.cc]: Rename it to `test_thread_pool` in this commit https://github.com/tensorflow/tensorflow/pull/27648/commits/9cc83b27b1833f721c69dc877ef285000070162c

**Review by jsimsa — APPROVED** [review]

**Comment by rthadur** [comment]
@feihugis can you please check failed build errors ?

**Review by jsimsa — APPROVED** [review]

**Comment by feihugis** [comment]
@rthadur The failures seem to be unrelated:

- `Linux GPU`: caused by `distribute:multi_worker_test_gpu`

- `Ubuntu contrib`: caused by `//tensorflow/contrib/cudnn_rnn:cudnn_rnn_ops_test`

- `Windows Bazel`: caused by `//tensorflow/cc:array_ops_genrule`

- `Windows Bazel GPU`: caused by `//tensorflow/contrib/resampler:python/ops/_resampler_ops_gpu`

The logs for `Android Demo App` and `MacOS Contrib` cannot be accessed.

**Comment by feihugis** [comment]
> @rthadur The failures seem to be unrelated:
> 
> * `Linux GPU`: caused by `distribute:multi_worker_test_gpu`
> * `Ubuntu contrib`: caused by `//tensorflow/contrib/cudnn_rnn:cudnn_rnn_ops_test`
> * `Windows Bazel`: caused by `//tensorflow/cc:array_ops_genrule`
> * `Windows Bazel GPU`: caused by `//tensorflow/contrib/resampler:python/ops/_resampler_ops_gpu`
> 
> The logs for `Android Demo App` and `MacOS Contrib` cannot be accessed.

The internal checks failed. Could you help paste the logs here?

**Comment by feihugis** [comment]
> @rthadur The failures seem to be unrelated:
> 
> * `Linux GPU`: caused by `distribute:multi_worker_test_gpu`
> * `Ubuntu contrib`: caused by `//tensorflow/contrib/cudnn_rnn:cudnn_rnn_ops_test`
> * `Windows Bazel`: caused by `//tensorflow/cc:array_ops_genrule`
> * `Windows Bazel GPU`: caused by `//tensorflow/contrib/resampler:python/ops/_resampler_ops_gpu`
> 
> The logs for `Android Demo App` and `MacOS Contrib` cannot be accessed.

Google internal checks FAILED. Could you help paste the logs here?

**Comment by jsimsa** [comment]
The failing tests are unrelated. @rthadur can you please re-trigger the internal submission.

**Pull request merged**